regulations/tasks.py

```diff
@@ -24,55 +25,71 @@
 
 from django.conf import settings
 from django.template import loader
+from django.utils.crypto import get_random_string
 
 from regulations.models import FailedCommentSubmission
 logger = get_task_logger(__name__)
 
 
 @shared_task(bind=True)
-def submit_comment(self, body):
+def submit_comment(self, comments, form_data, metadata_url):
     '''
     Submit the comment to regulations.gov. If unsuccessful, retry the task.
     Number of retries and time between retries is managed by Celery settings.
     The main comment is converted to a PDF and added as an attachment; the
     'general_comment' field refers to this attachment.
 
-    :param body: dict of fields and values to be posted to regulations.gov
+    :param comments: List of sectional comments
+    :param form_data: Dict of fields accepted by regulations.gov
+    :param metadata_url: SignedUrl for comment metadata
     '''
-    sections = body.get('assembled_comment', [])
     try:
-        html = json_to_html(sections)
-        files = extract_files(sections)
+        html = json_to_html(comments)
+        files = extract_files(comments)
         try:
             with html_to_pdf(html) as comment_pdf, \
                     build_attachments(files) as attachments:
-                data = build_multipart_encoded(body, comment_pdf, attachments)
+                pdf_url = cache_pdf(comment_pdf, metadata_url)
+                data = build_multipart_encoded(
+                    form_data, comment_pdf, attachments)
                 response = post_submission(data)
                 if response.status_code != requests.codes.created:
                     logger.warn("Post to regulations.gov failed: %s %s",
                                 response.status_code, response.text)
                     raise self.retry()
                 logger.info(response.text)
-                return response.json()
+                data = response.json()
+                return {
+                    'trackingNumber': data['trackingNumber'],
+                    'pdfUrl': pdf_url.url,
+                }
         except (RequestException, ClientError):
             logger.exception("Unable to submit comment")
             raise self.retry()
     except MaxRetriesExceededError:
         message = "Exceeded retries, saving failed submission"
         logger.error(message)
-        save_failed_submission(json.dumps(body))
-        return {'message': message, 'trackingNumber': None}
+        save_failed_submission(
+            json.dumps({'comments': comments, 'form_data': form_data})
+        )
+        return {
+            'pdfUrl': pdf_url.url,
+            'trackingNumber': None,
+        }
 
 
 @shared_task
-def publish_tracking_number(response, key):
+def publish_tracking_number(response, metadata_url):
     s3 = make_s3_client()
-    body = {'trackingNumber': response['trackingNumber']}
+    body = {
+        'pdfUrl': response['pdfUrl'],
+        'trackingNumber': response['trackingNumber'],
+    }
```

Comment:
Curious: is `response` just the return value from `submit_comment`? If so, would it make sense to skip the `body` here and just `json.dumps` the response?